File: py/candlestick_patterns/core/pattern_identifier.py
"""Enumeration of the 61 candlestick pattern identifiers."""
from __future__ import annotations

from enum import IntEnum


class PatternIdentifier(IntEnum):
    """Identifier for each of the 61 candlestick patterns.

    Values are sequential starting at 0, sorted alphabetically.
    The ``name`` attribute matches the method name on ``CandlestickPatterns``.
    """
    ABANDONED_BABY = 0
    ADVANCE_BLOCK = 1
    BELT_HOLD = 2
    BREAKAWAY = 3
    CLOSING_MARUBOZU = 4
    CONCEALING_BABY_SWALLOW = 5
    COUNTERATTACK = 6
    DARK_CLOUD_COVER = 7
    DOJI = 8
    DOJI_STAR = 9
    DRAGONFLY_DOJI = 10
    ENGULFING = 11
    EVENING_DOJI_STAR = 12
    EVENING_STAR = 13
    GRAVESTONE_DOJI = 14
    HAMMER = 15
    HANGING_MAN = 16
    HARAMI = 17
    HARAMI_CROSS = 18
    HIGH_WAVE = 19
    HIKKAKE = 20
    HIKKAKE_MODIFIED = 21
    HOMING_PIGEON = 22
    IDENTICAL_THREE_CROWS = 23
    IN_NECK = 24
    INVERTED_HAMMER = 25
    KICKING = 26
    KICKING_BY_LENGTH = 27
    LADDER_BOTTOM = 28
    LONG_LEGGED_DOJI = 29
    LONG_LINE = 30
    MARUBOZU = 31
    MATCHING_LOW = 32
    MAT_HOLD = 33
    MORNING_DOJI_STAR = 34
    MORNING_STAR = 35
    ON_NECK = 36
    PIERCING = 37
    RICKSHAW_MAN = 38
    RISING_FALLING_THREE_METHODS = 39
    SEPARATING_LINES = 40
    SHOOTING_STAR = 41
    SHORT_LINE = 42
    SPINNING_TOP = 43
    STALLED = 44
    STICK_SANDWICH = 45
    TAKURI = 46
    TASUKI_GAP = 47
    THREE_BLACK_CROWS = 48
    THREE_INSIDE = 49
    THREE_LINE_STRIKE = 50
    THREE_OUTSIDE = 51
    THREE_STARS_IN_THE_SOUTH = 52
    THREE_WHITE_SOLDIERS = 53
    THRUSTING = 54
    TRISTAR = 55
    TWO_CROWS = 56
    UNIQUE_THREE_RIVER = 57
    UP_DOWN_GAP_SIDE_BY_SIDE_WHITE_LINES = 58
    UPSIDE_GAP_TWO_CROWS = 59
    X_SIDE_GAP_THREE_METHODS = 60

    @property
    def method_name(self) -> str:
        """The snake_case method name on ``CandlestickPatterns``."""
        return self.name.lower()

File: go/candlestickpatterns/core/patternidentifier.go
package core

// PatternIdentifier identifies each of the 61 candlestick patterns.
// Values are sequential starting at 0, sorted alphabetically.
type PatternIdentifier int

const (
	AbandonedBaby                     PatternIdentifier = iota // 0
	AdvanceBlock                                               // 1
	BeltHold                                                   // 2
	Breakaway                                                  // 3
	ClosingMarubozu                                            // 4
	ConcealingBabySwallow                                      // 5
	Counterattack                                              // 6
	DarkCloudCover                                             // 7
	Doji                                                       // 8
	DojiStar                                                   // 9
	DragonflyDoji                                              // 10
	Engulfing                                                  // 11
	EveningDojiStar                                            // 12
	EveningStar                                                // 13
	GravestoneDoji                                             // 14
	Hammer                                                     // 15
	HangingMan                                                 // 16
	Harami                                                     // 17
	HaramiCross                                                // 18
	HighWave                                                   // 19
	Hikkake                                                    // 20
	HikkakeModified                                            // 21
	HomingPigeon                                               // 22
	IdenticalThreeCrows                                        // 23
	InNeck                                                     // 24
	InvertedHammer                                             // 25
	Kicking                                                    // 26
	KickingByLength                                            // 27
	LadderBottom                                               // 28
	LongLeggedDoji                                             // 29
	LongLine                                                   // 30
	Marubozu                                                   // 31
	MatchingLow                                                // 32
	MatHold                                                    // 33
	MorningDojiStar                                            // 34
	MorningStar                                                // 35
	OnNeck                                                     // 36
	Piercing                                                   // 37
	RickshawMan                                                // 38
	RisingFallingThreeMethods                                  // 39
	SeparatingLines                                            // 40
	ShootingStar                                               // 41
	ShortLine                                                  // 42
	SpinningTop                                                // 43
	Stalled                                                    // 44
	StickSandwich                                              // 45
	Takuri                                                     // 46
	TasukiGap                                                  // 47
	ThreeBlackCrows                                            // 48
	ThreeInside                                                // 49
	ThreeLineStrike                                            // 50
	ThreeOutside                                               // 51
	ThreeStarsInTheSouth                                       // 52
	ThreeWhiteSoldiers                                         // 53
	Thrusting                                                  // 54
	Tristar                                                    // 55
	TwoCrows                                                   // 56
	UniqueThreeRiver                                           // 57
	UpDownGapSideBySideWhiteLines                              // 58
	UpsideGapTwoCrows                                          // 59
	XSideGapThreeMethods                                       // 60
)

// PatternCount is the total number of patterns.
const PatternCount = 61

// MethodName returns the snake_case method name (matching the Python convention).
func (p PatternIdentifier) MethodName() string {
	if int(p) < len(patternNames) {
		return patternNames[p]
	}
	return ""
}

var patternNames = [PatternCount]string{
	"abandoned_baby",
	"advance_block",
	"belt_hold",
	"breakaway",
	"closing_marubozu",
	"concealing_baby_swallow",
	"counterattack",
	"dark_cloud_cover",
	"doji",
	"doji_star",
	"dragonfly_doji",
	"engulfing",
	"evening_doji_star",
	"evening_star",
	"gravestone_doji",
	"hammer",
	"hanging_man",
	"harami",
	"harami_cross",
	"high_wave",
	"hikkake",
	"hikkake_modified",
	"homing_pigeon",
	"identical_three_crows",
	"in_neck",
	"inverted_hammer",
	"kicking",
	"kicking_by_length",
	"ladder_bottom",
	"long_legged_doji",
	"long_line",
	"marubozu",
	"matching_low",
	"mat_hold",
	"morning_doji_star",
	"morning_star",
	"on_neck",
	"piercing",
	"rickshaw_man",
	"rising_falling_three_methods",
	"separating_lines",
	"shooting_star",
	"short_line",
	"spinning_top",
	"stalled",
	"stick_sandwich",
	"takuri",
	"tasuki_gap",
	"three_black_crows",
	"three_inside",
	"three_line_strike",
	"three_outside",
	"three_stars_in_the_south",
	"three_white_soldiers",
	"thrusting",
	"tristar",
	"two_crows",
	"unique_three_river",
	"up_down_gap_side_by_side_white_lines",
	"upside_gap_two_crows",
	"x_side_gap_three_methods",
}
